### PR TITLE
[TODO] Remove some boxing from IPv6AddressHelper.CreateCanonicalName

### PR DESCRIPTION
**Pull request by jamesqo**
Don't use `StringBuilder.AppendFormat` which converts the rest of the arguments to objects, instead call the `ToString` overload accepting a format/format provider ourselves and `Append` it to the builder.

cc @stephentoub @davidsh

### CODE REVIEW

File: src/System.Private.Uri/src/System/IPv6AddressHelper.cs [hunk 1]
```diff
@@ -44,13 +41,27 @@ internal unsafe static string CreateCanonicalName(ushort* numbers)
...
-                        numbers[i] >> 8, numbers[i] & 0xFF, numbers[i + 1] >> 8, numbers[i + 1] & 0xFF);
+
+                    // Call all of this manually, instead of using AppendFormat,
+                    // to avoid boxing
+                    // Equivalent to the format ":{0:d}.{1:d}.{2:d}.{3:d}"
+
+                    var first = (numbers[i] >> 8).ToString("d", formatProvider);
+                    var second = (numbers[i] & 0xFF).ToString("d", formatProvider);
+                    var third = (numbers[i + 1] >> 8).ToString("d", formatProvider);
+                    var fourth = (numbers[i + 1] & 0xFF).ToString("d", formatProvider);
```
Comment (stephentoub) [inline src/System.Private.Uri/src/System/IPv6AddressHelper.cs]: Nit: please use string instead of var here

**Comment by stephentoub** [comment]
This is one of those changes where it definitely improves things (there is significantly less memory allocated in the new version), but it really shouldn't be necessary.  A better implementation of StringBuilder.AppendFormat could alleviate the costs, at which point the older implementation would be better IMO because it's simpler and easier to read.

That said, I don't know that we'll see AppendFormat improved any time soon... so LGTM.

**Comment by davidsh** [comment]
cc: @CIPop

File: src/System.Private.Uri/src/System/IPv6AddressHelper.cs [hunk 2]
```diff
@@ -44,13 +41,27 @@ internal unsafe static string CreateCanonicalName(ushort* numbers)
...
+
             for (int i = 0; i < NumberOfLabels; i++)
             {
                 if (ipv4Embedded && i == (NumberOfLabels - 2))
                 {
                     // Write the remaining digits as an IPv4 address
-                    builder.AppendFormat(CultureInfo.InvariantCulture, EmbeddedIPv4Format,
-                        numbers[i] >> 8, numbers[i] & 0xFF, numbers[i + 1] >> 8, numbers[i + 1] & 0xFF);
+
+                    // Call all of this manually, instead of using AppendFormat,
```
Comment (CIPop) [inline src/System.Private.Uri/src/System/IPv6AddressHelper.cs]: Nit: end comments in "."

File: src/System.Private.Uri/src/System/IPv6AddressHelper.cs [hunk 3]
```diff
@@ -73,7 +84,7 @@ internal unsafe static string CreateCanonicalName(ushort* numbers)
                 {
                     builder.Append(Separator);
                 }
-                builder.AppendFormat(CultureInfo.InvariantCulture, CanonicalNumberFormat, numbers[i]);
+                builder.Append(numbers[i].ToString("x", formatProvider)); // Lower case hex, no leading zeros
```
Comment (CIPop) [inline src/System.Private.Uri/src/System/IPv6AddressHelper.cs]: Nit: end comment in "."

**Comment by CIPop** [comment]
@jamesqo please add tests to prove that this is making it better and to ensure it is not regressing in the future.

**Comment by stephentoub** [comment]
@jamesqo, are you still working on this? Thanks.

**Comment by jamesqo** [comment]
@stephentoub Sorry, I had forgotten about this PR... was working on setting up the project file for the performance tests a while ago as @CIPop suggested, but I switched to working on another PR at that time. I'm going to push the performance tests sometime soon.

**Comment by stephentoub** [comment]
Sounds good.  Thanks, @jamesqo.

**Comment by stephentoub** [comment]
@jamesqo, still in progress?

**Comment by jamesqo** [comment]
@stephentoub Sorry for keeping this open for so long. I'll close this + reopen once the perf tests are ready.

**Comment by stephentoub** [comment]
Ok, thanks.

**Pull request closed without merging**